### PR TITLE
Ampers - Abinnet

### PR DESCRIPTION
**Pull request by Abiaina**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | Deciding to make a room class or just an array of room ids. Talking it over with Chris during our on-on-one and in class when I recognized that the room class is just holding very little info.
Describe a concept that you gained more clarity on as you worked on this assignment. | I improved on design, I used more flow charts/ mapping to come up with a basic design, then wrote the tests and continuously ran the tests as I wrote the code itself.
Describe a _nominal test_ that you wrote for this assignment. |  testing if a reservation is able to be made when dates are given and there is available rooms
Describe an _edge case test_ that you wrote for this assignment. | Testing if a reservation is able to be made when dates are given but there are no more available rooms. Trying to book more than 20 rooms in a single date range.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I actually liked it, the pseudo code part on whiteboards was really helpful in determining edge cases and what variables/parameters I would use for specific methods.

### CODE REVIEW

File: lib/admin.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  end
+
+  def add_reservation(check_in, check_out, room_id)
+    available_rooms = available_rooms(check_in, check_out)
+
+    if !(available_rooms.include?(room_id))
+      raise ArgumentError.new("#{room_id} is booked for these dates")
+    end
+
+    if available_rooms.count == 0
```
Comment (CheezItMan) [inline lib/admin.rb:27]: Ummm.. if there were no rooms then `available_rooms` would not include `room_id`.

File: lib/admin.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  def available_rooms(check_in, check_out)
+    check_in = Date.parse(check_in)
+    check_out = Date.parse(check_out)
+
+    rooms = @roomlist.dup
+
+    @reservations.each do |reservation|
+      valid_check_in = check_in.between?(reservation.start_date, (reservation.end_date - 1))
+
+      valid_check_out = check_out.between?((reservation.start_date + 1), reservation.end_date)
```
Comment (CheezItMan) [inline lib/admin.rb]: What about if the reservation occurs entirely between `check_in` and `check_out`?

File: lib/admin.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    end
+
+    if available_rooms.count == 0
+      raise ArgumentError.new("No available roooms")
+    end
+
+    @reservations << Reservation.new(check_in, check_out, room_id)
+  end
+
+  def available_rooms(check_in, check_out)
```
Comment (CheezItMan) [inline lib/admin.rb]: To have wave 3 complete you may also need to check the blocks, or make sure `add_reservation` considers blocks.

File: lib/reservation.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
+require 'date'
+
+class Reservation
```
Comment (CheezItMan) [inline lib/reservation.rb:3]: I would suggest a method in reservation which either takes a `Reservation` or a date range and returns true if it overlaps with this reservation and false if it does not.  It might make the logic easier in Admin.

File: specs/spec_admin.rb [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+  it "reservations returns list of reservation instances" do
+    @bookings.must_be_instance_of Array
+    @bookings[0].must_be_instance_of Reservation
+  end
+
+  it "can access room id from reservation instance." do
+    @bookings[0].room_id.must_equal 1
+  end
+
+  it "calculates the length of stay" do
```
Comment (CheezItMan) [inline specs/spec_admin.rb:41]: This is really a test of the `Reservation` class.

File: specs/spec_block.rb [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    @blockroom_test.must_be_instance_of BlockRoom
+  end
+
+  it "check_in/check_out instance variables are of date class" do
+    @blockroom_test.check_in.must_be_instance_of Date
+    @blockroom_test.check_out.must_be_instance_of Date
+  end
+
+  it "add reservation adds reservation to reservations list" do
+    before_rezs =  @blockroom_test.reservations.dup
```
Comment (CheezItMan) [inline specs/spec_block.rb:25]: I'm unsure why you need to duplicate the array.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Design** |
Demonstrated classes having a single responsibility |  For the most part, your `Admin` class is doing a bit much, but overall pretty good.
Demonstrated loose coupling |  Very good work here, `Admin` depends on `Reservation` and `Block`, and `Block` depends on `Reservation`
Methods demonstrate a good use of encapsulation, inputs and outputs | Pretty good, see my comments.
Wave 1 requirements |  Check
Wave 2 requirements |  Check, but you have a bug in your logic, see my notes. 
Wave 3 requirements |  Check
Summary | Nice work, you showed a lot of growth here in design.  Overall nicely done.  You have some bugs in your code, but the design is pretty good overall.  Checkout our [reference implementation](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/02-intermediate-ruby/exercises/hotel-revisited.md)
